code/projects/steg/src/lib/src/vc.rs

```diff
@@ -0,0 +1,202 @@
+use crate::image_data::{create_image_data, ImageDataExt, RGBA_CHANNELS};
+use tinyrand::{Rand, Seeded, StdRand};
+use wasm_bindgen::prelude::*;
+use web_sys::ImageData;
+
+const BLACK: u8 = u8::MIN;
+const WHITE: u8 = u8::MAX;
+
+#[wasm_bindgen]
+#[derive(Clone, Copy, Debug)]
+pub enum VcGrayscale {
+    Average = "average",
+    Luminosity = "luminosity",
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Copy, Debug)]
+pub enum VcDithering {
+    Threshold = "threshold",
+    Random = "random",
+    FloydSteinberg = "floyd-steinberg",
+}
+
+fn closest_palette_color(v: u8) -> u8 {
+    if v < u8::MAX / 2 {
+        BLACK
+    } else {
+        WHITE
+    }
+}
+
+fn pixels_to_image(pixels: Vec<Vec<u8>>) -> ImageData {
+    let w = pixels.first().map(|p| p.len()).unwrap_or_default();
+    let h = pixels.len();
+    let mut data = vec![u8::MAX; w * h * RGBA_CHANNELS];
+    for i in 0..(data.len() / RGBA_CHANNELS) {
+        let v = pixels[i / w][i % w];
+        let i = i * RGBA_CHANNELS;
+        data[i] = v;
+        data[i + 1] = v;
+        data[i + 2] = v;
+    }
+    create_image_data(&data, w as u32)
+}
+
+fn desaturate(image: &ImageData, grayscale: VcGrayscale) -> ImageData {
+    image.map_pixels(|(r, g, b)| {
+        let v = match grayscale {
+            VcGrayscale::Average => 0.333 * r as f64 + 0.333 * g as f64 + 0.333 * b as f64,
+            VcGrayscale::Luminosity => 0.299 * r as f64 + 0.587 * g as f64 + 0.114 * b as f64,
```

Comment:
Average: Adding `r + g + b` and then dividing them in 3 could be better. We are loosing a little bit of precision for every 0.333 division.
Luminosity: Constants for `YUV` luminosity could be useful.

Reply:
Fixed locally, I'll push it to the remote soon.